### ISSUE TITLE
Check correctness of CLiMF

### ISSUE THREAD

**Issue by salvacarrion** [biolab/orange3-recommendation#1]
**CLiMF pass all the tests**, so functions _MRR_ and _loss_ are maximized successfully but the increase on their values per iteration is too small (even for higher learning rates).

**Checks**:
1. Functions _MRR_ and _loss_ have been check against other libraries providing the same results for the same set of inputs. Hence, we can presume that they are correct.
2. `f = np.einsum('j,ij->i', U[i], V[items])` provide the same results as `f = precompute_f(data,U,V,i)` but will half precision (I've tried to double it by using _np.float64/np.float128_ without success ). Maybe this could be the cause.

**Run test:**

```
# Function in test_climf.py
def test_mrr(self):
    learner = CLiMFLearner(num_factors=10, num_iter=10, verbose=3, learning_rate=0.0001, lmbda=0.001)
    super().test_mrr(learner, filename='epinions_train.tab', testdata='epinions_test.tab')
```

**Output:**

```
CLiMF factorisation started.
    Level of verbosity: 3
        - Verbosity = 1 ->  [time/iter]
        - Verbosity = 2 ->  [time/iter, loss]
        - Verbosity = 3 ->  [time/iter, loss, MRR]

- Step: 1
    - Time: 1.383s
    - Training loss: -98104.785
    - Train MRR: 0.0532

- Step: 2
    - Time: 1.536s
    - Training loss: -98104.462
    - Train MRR: 0.0657

- Step: 3
    - Time: 1.367s
    - Training loss: -98104.140
    - Train MRR: 0.0685

- Step: 4
    - Time: 1.415s
    - Training loss: -98103.817
    - Train MRR: 0.0698

- Step: 5
    - Time: 1.307s
    - Training loss: -98103.494
    - Train MRR: 0.0698

- Step: 6
    - Time: 1.287s
    - Training loss: -98103.171
    - Train MRR: 0.0696

- Step: 7
    - Time: 1.321s
    - Training loss: -98102.847
    - Train MRR: 0.0693

- Step: 8
    - Time: 1.299s
    - Training loss: -98102.522
    - Train MRR: 0.0697

- Step: 9
    - Time: 1.365s
    - Training loss: -98102.198
    - Train MRR: 0.0693

- Step: 10
    - Time: 1.301s
    - Training loss: -98101.872
    - Train MRR: 0.0694

MRR: 0.0856
```

**Output should be similar to this:**

```
iteration 1:
train mrr = 0.05750323
iteration 2:
train mrr = 0.06994045
iteration 3:
train mrr = 0.07469046
iteration 4:
train mrr = 0.07611490
iteration 5:
train mrr = 0.07690448
iteration 6:
train mrr = 0.07711117
iteration 7:
train mrr = 0.07657648
iteration 8:
train mrr = 0.07625065
iteration 9:
train mrr = 0.07590930
iteration 10:
train mrr = 0.07558414

Test MRR: Test MRR: 0.40167283
```

**Relevant links:**
- **Pure python code**: https://github.com/gamboviol/climf
- **Cython code**: https://github.com/coreylynch/pyCLiMF

**Comment by salvacarrion** [comment]
Problem solved.

Finally, there was no problem with the implementation. The test file was corrupted.

```
CLiMF factorization started.
    Level of verbosity: 3
        - Verbosity = 1 ->  [time/iter]
        - Verbosity = 2 ->  [time/iter, loss]
        - Verbosity = 3 ->  [time/iter, loss, MRR]

- Step: 1
    - Time: 1.332s
    - Training loss: -98104.766
    - Train MRR: 0.0558

- Step: 2
    - Time: 1.367s
    - Training loss: -98104.443
    - Train MRR: 0.0704

- Step: 3
    - Time: 1.315s
    - Training loss: -98104.119
    - Train MRR: 0.0730

- Step: 4
    - Time: 1.320s
    - Training loss: -98103.795
    - Train MRR: 0.0739

- Step: 5
    - Time: 1.325s
    - Training loss: -98103.471
    - Train MRR: 0.0747

- Step: 6
    - Time: 1.328s
    - Training loss: -98103.146
    - Train MRR: 0.0752

- Step: 7
    - Time: 1.333s
    - Training loss: -98102.821
    - Train MRR: 0.0750

- Step: 8
    - Time: 1.309s
    - Training loss: -98102.495
    - Train MRR: 0.0756

- Step: 9
    - Time: 1.306s
    - Training loss: -98102.169
    - Train MRR: 0.0758

- Step: 10
    - Time: 1.341s
    - Training loss: -98101.843
    - Train MRR: 0.0757

MRR: 0.3975
```

**Comment by yaronhe** [comment]
The update equation is different from the original paper 
dV = _g(-f[j]) - beta * V[w]
dV += np.einsum('i,j->ij', vec1, U[i]).sum(axis=0)

It seems that the first equation should 
dV = _g(-f[j])  **U_i** - beta * V[w]
